d2layouts/d2sequence/sequence_diagram.go

```diff
@@ -0,0 +1,312 @@
+		minChildY := math.Inf(1)
+		maxChildY := math.Inf(-1)
+		for _, child := range span.ChildrenArray {
+			minChildY = math.Min(minChildY, child.TopLeft.Y)
+			maxChildY = math.Max(maxChildY, child.TopLeft.Y+child.Height)
+		}
+
+		// finds the position if there are messages to this span
+		minMessageY := math.Inf(1)
+		if minRank, exists := sd.minMessageRank[span]; exists {
+			minMessageY = sd.getMessageY(minRank)
+		}
+		maxMessageY := math.Inf(-1)
+		if maxRank, exists := sd.maxMessageRank[span]; exists {
+			maxMessageY = sd.getMessageY(maxRank)
+		}
+
+		// if it is the same as the child top left, add some padding
+		minY := math.Min(minMessageY, minChildY)
+		if minY == minChildY {
+			minY -= SPAN_DEPTH_GROW_FACTOR
+		} else {
+			minY -= SPAN_MESSAGE_PAD
+		}
+		maxY := math.Max(maxMessageY, maxChildY)
+		if maxY == maxChildY {
+			maxY += SPAN_DEPTH_GROW_FACTOR
+		} else {
+			maxY += SPAN_MESSAGE_PAD
+		}
+
+		height := math.Max(maxY-minY, MIN_SPAN_HEIGHT)
+		// -2 because the actors count as level 1 making the first level span getting 2*SPAN_DEPTH_GROW_FACTOR
+		width := SPAN_WIDTH + (float64(span.Level()-2) * SPAN_DEPTH_GROW_FACTOR)
+		x := rankToX[sd.objectRank[span]] - (width / 2.)
+		span.Box = geo.NewBox(geo.NewPoint(x, minY), width, height)
+		span.ZIndex = 1
+	}
+}
+
+// routeMessages routes horizontal edges (messages) from Src to Dst
+func (sd *sequenceDiagram) routeMessages() {
+	for rank, message := range sd.messages {
+		isLeftToRight := message.Src.TopLeft.X < message.Dst.TopLeft.X
+
+		// finds the proper anchor point based on the message direction
+		var startX, endX float64
+		if sd.isActor(message.Src) {
+			startX = message.Src.Center().X
+		} else if isLeftToRight {
+			startX = message.Src.TopLeft.X + message.Src.Width
+		} else {
+			startX = message.Src.TopLeft.X
+		}
+
+		if sd.isActor(message.Dst) {
+			endX = message.Dst.Center().X
+		} else if isLeftToRight {
+			endX = message.Dst.TopLeft.X
+		} else {
+			endX = message.Dst.TopLeft.X + message.Dst.Width
+		}
+
+		if isLeftToRight {
+			startX += SPAN_MESSAGE_PAD
+			endX -= SPAN_MESSAGE_PAD
+		} else {
+			startX -= SPAN_MESSAGE_PAD
+			endX += SPAN_MESSAGE_PAD
+		}
+
+		messageY := sd.getMessageY(rank)
+		message.Route = []*geo.Point{
+			geo.NewPoint(startX, messageY),
+			geo.NewPoint(endX, messageY),
+		}
+
+		if message.Attributes.Label.Value != "" {
+			if isLeftToRight {
+				message.LabelPosition = go2.Pointer(string(label.OutsideTopCenter))
+			} else {
+				// the label will be placed above the message because the orientation is based on the edge normal vector
+				message.LabelPosition = go2.Pointer(string(label.OutsideBottomCenter))
+			}
+		}
+	}
+}
+
+func (sd *sequenceDiagram) getMessageY(rank int) float64 {
+	// +1 so that the first message has the top padding for its label
+	return ((float64(rank) + 1.) * sd.messageYStep) + sd.maxActorHeight
+}
+
+func (sd *sequenceDiagram) isActor(obj *d2graph.Object) bool {
```

Comment:
save the root object on the sequenceDiagram. Then, if the obj.parent == the root object, then u know it's an actor